pkg/capabilities/consensus/ocr3.go

```diff
@@ -0,0 +1,110 @@
+package consensus
+
+import (
+	"context"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/shopspring/decimal"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+)
+
+var info = capabilities.MustNewCapabilityInfo(
+	"ocr3",
+	capabilities.CapabilityTypeConsensus,
+	"OCR3 consensus exposed as a capability.",
+	"v1.0.0",
+)
+
+type cache struct {
+	requests        map[string]capabilities.CapabilityRequest
+	responses       map[string]capabilities.CapabilityResponse
+	responseChannel chan capabilities.CapabilityResponse
+}
+
+func newCache() *cache {
+	return &cache{
+		requests:        map[string]capabilities.CapabilityRequest{},
+		responses:       map[string]capabilities.CapabilityResponse{},
+		responseChannel: make(chan capabilities.CapabilityResponse),
+	}
+}
+
+type ocr3Store struct {
+	workflowExecutionIDs []string
+	cache                *cache
+}
+
+func newOCR3Store() *ocr3Store {
+	return &ocr3Store{
+		workflowExecutionIDs: []string{},
+		cache:                newCache(),
+	}
+}
+
+func (o *ocr3Store) addCapabilityRequest(request capabilities.CapabilityRequest) {
+	o.workflowExecutionIDs = append(o.workflowExecutionIDs, request.Metadata.WorkflowExecutionID)
+	o.cache.requests[request.Metadata.WorkflowExecutionID] = request
+}
+
+type OCR3 struct {
+	capabilities.CapabilityInfo
+	capabilities.CallbackExecutable
+	ocr3Store *ocr3Store
+}
+
+func NewOCR3() *OCR3 {
+	return &OCR3{
+		CapabilityInfo: info,
+		ocr3Store:      newOCR3Store(),
+	}
+}
+
+func (o *OCR3) RegisterToWorkflow(ctx context.Context, request capabilities.RegisterToWorkflowRequest) error {
+	return nil
+}
+
+func (o *OCR3) UnregisterFromWorkflow(ctx context.Context, request capabilities.UnregisterFromWorkflowRequest) error {
+	return nil
+}
+
+type CapabilityInputs struct {
+	Observations map[string]decimal.Decimal
+}
+
+func (o *OCR3) Execute(ctx context.Context, callback chan<- capabilities.CapabilityResponse, request capabilities.CapabilityRequest) error {
+	// Receives and stores an observation to do consensus on
+	// Receives an aggregation method; at this point the method has been validated
+	// Returns the consensus result over a channel
+	o.ocr3Store.addCapabilityRequest(request)
+
+	valuesMap, err := request.Inputs.Unwrap()
+	if err != nil {
+		return err
+	}
+
+	capabilityInputs := &CapabilityInputs{}
+	err = mapstructure.Decode(valuesMap, capabilityInputs)
+	if err != nil {
+		return err
+	}
+
+	// --- START OF MOCKED OCR3 LOGIC ---
```

Comment:
Our plan was to implement both sides of the interface in the same struct -- there's no point keeping them separate since they aren't reusable separately.

I don't have a strong feeling about this though, so happy to go with separating them.

Reply:
Don't we need to have something to pass to ReportingPluginFactory so that the plugin can use it? I imagine that libocr's Oracle is a separate service from the one that handles Capability Interface calls...? Or were you just planning to use a struct for that (if so, having an interface makes it easier to mock stuff)...?